### PR TITLE
expression/types: fix decimal minus/round/multiple result

### LINKED ISSUE #6997: negative small decimal return unexcept result
this issue relates to #6985 but is not the only one question source.

1. What did you do?

```
mysql> select -0.0000000000000000000000000000000000000000000000000017382578996420603;
```

2. What did you expect to see?

```
Field   1:  `-0.0000000000000000000000000000000000000000000000000017382578996420603`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     71
Max_length: 70
Decimals:   67
Flags:      NOT_NULL BINARY NUM 


+------------------------------------------------------------------------+
| -0.0000000000000000000000000000000000000000000000000017382578996420603 |
+------------------------------------------------------------------------+
| -0.0000000000000000000000000000000000000000000000000017382578996420603 |
+------------------------------------------------------------------------+

```

3. What did you see instead?

```
Field   1:  `-0.0000000000000000000000000000000000000000000000000017382578996420603`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     72
Max_length: 32
Decimals:   67
Flags:      BINARY NUM 


+------------------------------------------------------------------------+
| -0.0000000000000000000000000000000000000000000000000017382578996420603 |
+------------------------------------------------------------------------+
|                                       0.000000000000000000000000000000 |
+------------------------------------------------------------------------+

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v2.1.0-beta-11-g1f6e12cc5
Git Commit Hash: 1f6e12cc55e534bf514eef212f1091a91b8592c9
Git Branch: HEAD
UTC Build Time: 2018-07-04 06:31:26
GoVersion: go version go1.10.1 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
```

### PR DESCRIPTION
**Pull request by lysu**
## What have you changed? (mandatory)

this PR mainly fix three questions:

- minus sign(`-x`) direct change decimal's negative property. (relate to #6997)
- change scale truncate action in `decimal.round` base on the result of expression's decimal type infer 
- fix `decimal.round` expression's decimals type infer which help 2nd fix
- fix multiple decimal panic (relate to #6985 as 3deb8f5b7c9f54bc1a9a4df979b93a1c13a9c213) 
- ignore data truncated error for multiple operation 
- make pingcap/tidb-test#545 pass

fixes #6997, #6985

## What are the type of the changes (mandatory)?

The currently defined types are listed below, please pick one of the types for this PR by removing the others:

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested (mandatory)?

- unit tests
- integration tests(pingcap/tidb-test#545)

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7001)

### CODE REVIEW

File: server/util.go [hunk 1]
```diff
@@ -254,7 +254,7 @@ func dumpBinaryRow(buffer []byte, columns []*ColumnInfo, row types.Row) ([]byte,
 		case mysql.TypeDouble:
 			buffer = dumpUint64(buffer, math.Float64bits(row.GetFloat64(i)))
 		case mysql.TypeNewDecimal:
-			buffer = dumpLengthEncodedString(buffer, hack.Slice(row.GetMyDecimal(i).String()))
+			buffer = dumpLengthEncodedString(buffer, hack.Slice(string(row.GetMyDecimal(i).ToString())))
```
Comment (shenli) [inline server/util.go]: Why change this?

Comment (lysu) [reply server/util.go]: mmm...In MySQL `select -0.0000000000000000000000000000000000000000000000000017382578996420603` will not round up to zero...so I try to run test case ahead in here to see what happen if we don't truncate it- -

File: types/mydecimal.go [hunk 2]
```diff
@@ -1383,6 +1383,20 @@ func (d *MyDecimal) Compare(to *MyDecimal) int {
 	return 1
 }
 
+// DecimalNeg reverses decimal's sign.
+func DecimalNeg(from *MyDecimal) *MyDecimal {
+	to := *from
+	if from.IsZero() {
+		return &to
+	}
+	if from.IsNegative() {
```
Comment (coocood) [inline types/mydecimal.go]: How about `to.negative = !from.negative`?

File: server/util.go (hunk 1 above)
Comment (crazycs520) [inline server/util.go]: buffer = dumpLengthEncodedString(buffer, row.GetMyDecimal(i).ToString())

File: server/util.go [hunk 3]
```diff
@@ -312,7 +312,7 @@ func dumpTextRow(buffer []byte, columns []*ColumnInfo, row types.Row) ([]byte, e
 			tmp = appendFormatFloat(tmp[:0], row.GetFloat64(i), prec, 64)
 			buffer = dumpLengthEncodedString(buffer, tmp)
 		case mysql.TypeNewDecimal:
-			buffer = dumpLengthEncodedString(buffer, hack.Slice(row.GetMyDecimal(i).String()))
+			buffer = dumpLengthEncodedString(buffer, hack.Slice(string(row.GetMyDecimal(i).ToString())))
```
Comment (crazycs520) [inline server/util.go]: ditto

File: expression/builtin_math.go [hunk 4]
```diff
@@ -1768,7 +1769,7 @@ func (b *builtinTruncateDecimalSig) evalDecimal(row types.Row) (*types.MyDecimal
 	}
 
 	result := new(types.MyDecimal)
-	if err := x.Round(result, int(d), types.ModeTruncate); err != nil {
+	if err := x.Round(result, mathutil.Min(int(d), b.getRetTp().Decimal), types.ModeTruncate); err != nil {
```
Comment (lysu) [inline expression/builtin_math.go]: Truncate's param is given by user, so need check dec < tp.Decimal(which is less than MAX_DECIMAL_SCALE)

e.g. `Truncate(1.124, 100)`(we had test case).

File: types/mydecimal.go (hunk 2 above)
Comment (lysu) [reply types/mydecimal.go]: done~

File: server/util.go (hunk 1 above)
Comment (lysu) [reply server/util.go]: good idea~finally we revert it and use continue call round up.- -

File: types/mydecimal.go [hunk 5]
```diff
@@ -713,9 +713,6 @@ func (d *MyDecimal) doMiniRightShift(shift, beg, end int) {
 // RETURN VALUE
 //  eDecOK/eDecTruncated
 func (d *MyDecimal) Round(to *MyDecimal, frac int, roundMode RoundMode) (err error) {
-	if frac > mysql.MaxDecimalScale {
-		frac = mysql.MaxDecimalScale
-	}
```
Comment (lysu) [inline types/mydecimal.go]: https://github.com/mysql/mysql-server/blob/ba1a99c5cd7dd6204abbe8d177c7f71a48e53fe3/strings/decimal.c#L1641 In port source doesn't check it too.

**Comment by lysu** [comment]
run-common-test tidb-test=pr/545
run-integration-common-test=pr/545

**Comment by lysu** [comment]
/run-common-test tidb-test=pr/545
/run-integration-common-test=pr/545

File: expression/builtin_arithmetic.go [hunk 6]
```diff
@@ -523,7 +524,7 @@ func (s *builtinArithmeticMultiplyDecimalSig) evalDecimal(row types.Row) (*types
 	}
 	c := &types.MyDecimal{}
 	err = types.DecimalMul(a, b, c)
-	if err != nil {
+	if err != nil && !terror.ErrorEqual(err, types.ErrTruncated) {
```
Comment (lysu) [inline expression/builtin_arithmetic.go]:
```
  conventions:
    decimal_smth() <= 1     -- result is usable, but precision loss is possible
```
In [MySQL]( https://github.com/mysql/mysql-server/blob/220d2eb9623594f3bf3c81cf4f8af45c210c233b/include/decimal.h#L116) multiple will ignore mul DataTruncated error without any error or warning.....

**Comment by lysu** [comment]
PTAL @coocood @AndreMouche if free~ thx

File: types/mydecimal_test.go [hunk 7]
```diff
@@ -626,7 +645,7 @@ func (s *testMyDecimalSuite) TestMul(c *C) {
 		{"123456", "987654321", "121931851853376", nil},
 		{"123456", "9876543210", "1219318518533760", nil},
 		{"123", "0.01", "1.23", nil},
-		{"123", "0", "0", nil},
```
Comment (coocood) [inline types/mydecimal_test.go]: Why remove the old test case?

Comment (lysu) [reply types/mydecimal_test.go]: It's my mistake.- -

File: expression/builtin_arithmetic.go (hunk 6 above)
Comment (zz-jason) [reply expression/builtin_arithmetic.go]: Do we have some test to cover this behavior?

**Comment by coocood** [comment]
LGTM

**Comment by lysu** [comment]
PTAL @AndreMouche  @XuHuaiyu  if free, thx~

File: expression/builtin_math.go [hunk 8]
```diff
@@ -292,6 +295,25 @@ func (c *roundFunctionClass) getFunction(ctx sessionctx.Context, args []Expressi
 	return sig, nil
 }
 
+// fixDecimal4RoundAndTruncate fixes tp.decimals of round/truncate func.
+func fixDecimal4RoundAndTruncate(ctx sessionctx.Context, args []Expression, retType types.EvalType) int {
```
Comment (zz-jason) [inline expression/builtin_math.go]: s/fix.../calc.../

File: expression/builtin_math.go [hunk 9]
```diff
@@ -292,6 +295,25 @@ func (c *roundFunctionClass) getFunction(ctx sessionctx.Context, args []Expressi
...
+	if retType == types.ETInt || len(args) <= 1 {
+		return 0
+	}
+	secondConst, secondIsConst := args[1].(*Constant)
+	if !secondIsConst {
+		return args[0].GetType().Decimal
+	}
+	argDec, isNull, err := secondConst.EvalInt(ctx, nil)
+	if isNull || err != nil || argDec < 0 {
```
Comment (zz-jason) [inline expression/builtin_math.go]: we should check `err` firstly

**Pull request merged**